refactoring into submodules

choldgraf opened the pull request:
This refactors jupyter_sphinx into a few submodules:

* `ast.py`
* `execute.py`
* `thebelab.py`
* `utils.py`

and puts the sphinx `setup()` function in `__init__.py`.

the docs build and the tests pass for me. I still want to clean up the imports, and apply a `black` pass, but wanted to see if this looks reasonable to others first. If folks are +1 on this, then I'll add another commit that cleans up the import statements, and applies "black" and ensures a flake8 test passes.

The one user-facing change is that the extension will now be loaded with `jupyter_sphinx` instead of `jupyter_sphinx.execute`. Perhaps there's a way we could import one into the other to temporarily keep both behaviors?

akhmerov commented:
Wouldn't an `execute` module that's a complete copy of `__init__` (perhaps with a deprecation warning) not solve it?

Otherwise backwards compatibility breaking would also be OK, but then I'd like to go sure that also the directives and the AST are stabilized before making a release.

choldgraf commented:
ah good call - I just added a commit that is a little bit hacky, but I think accomplishes the same thing. Also added a deprecation warning. WDYT?

I'm not sure how to test *both* the old and new way...

choldgraf commented:
I just cleaned up the imports a bit (so they are more well-ordered) and ran `black` on the codebase. @akhmerov what do you think?

akhmerov commented on jupyter_sphinx/__init__.py:
```diff
@@ -1 +1,256 @@
+    app.add_config_value("jupyter_sphinx_continue_linenos", False, "env")
+
+    # Used for nodes that do not need to be rendered
+    def skip(self, node):
```
Here and later: wouldn't it better to factor the inline functions and constants out of `_setup`?

choldgraf replied:
good point - latest push has this in it

akhmerov commented on jupyter_sphinx/__init__.py:
```diff
@@ -1 +1,256 @@
+    return {"version": __version__, "parallel_read_safe": True}
+
+
+def setup(app):
```
This seems to be identical to `setup = _setup`; are you sure that we cannot just call `__init__.setup` from `execute`?

choldgraf replied:
ah yeah - I just refactored a bit more so something like this is done

akhmerov replied:
Thanks!

akhmerov commented on jupyter_sphinx/execute.py:
```diff
@@ -87,16 +59,16 @@ class JupyterKernel(Directive):
+        ]
 
 
 class JupyterKernelNode(docutils.nodes.Element):
```
Shouldn't the nodes be in `ast`?

choldgraf replied:
I was torn on this one - on one hand they're clearly AST objects, on the other they are explicitly about executing things. Do you have a preference? I'm happy to go with either

akhmerov replied:
I don't have a strong preference neither, but I think like putting all nodes (including this one) would make codebase a bit easier to navigate. Therefore I suggest indeed migrating this to `ast`. It's not a big point, and I'm happy with it either way.

choldgraf replied:
done in the latest push!

akhmerov commented on jupyter_sphinx/ast.py:
```diff
@@ -0,0 +1,351 @@
+import ipywidgets.embed
+import nbconvert
+
+import jupyter_sphinx as js
```
It's a cosmetic pont, but I prefer to not introduce abbreviations like this, except for the very standard `np` or `plt`.

At the same time, seeing where this is used, it seems that

```suggestion
from jupyter_sphinx import logger
```

would work even better.

choldgraf replied:
ah - the reason I was doing this is because the logger gets initialized by Sphinx so it's not importable at first. I'll lazily import the logger as you show above but inside the run function

akhmerov replied:
Ah, thanks for the explanation.

akhmerov commented:
Thanks a lot!

choldgraf commented:
woo - OK I'll try and upstream some more stuff this week
